forms/session/authentication_form.go

```diff
@@ -0,0 +1,64 @@
+package forms
+
+import (
+	"fmt"
+
+	"go-crawler-challenge/helpers"
+	"go-crawler-challenge/models"
+
+	log "github.com/beego/beego/v2/core/logs"
+	"github.com/beego/beego/v2/core/validation"
+)
+
+type AuthenticationForm struct {
+	Email    string `form:"email" valid:"Required; Email; MaxSize(100)"`
+	Password string `form:"password" valid:"Required;"`
+}
+
+var ValidationMessages = map[string]string{
+	"InvalidCredential": "Your email or password is incorrect",
+}
+
+var currentUser *models.User
+
+// Valid handles some custom form validation about validating existing user and checking match password
+func (form *AuthenticationForm) Valid(validation *validation.Validation) {
+	user, _ := models.GetUserByEmail(form.Email)
+	if user == nil {
+		err := validation.SetError("Email", ValidationMessages["InvalidCredential"])
+		if err != nil {
+			log.Warning(fmt.Sprintf("Set validation error failed: %v", err))
+		}
```

Comment:
@Lahphim Not sure that we should log a warning here since the `SerError` function does not return the error when it failed to set an error but it always returns the error which set to the validation.
```
// SetError Set error message for one field in ValidationError
func (v *Validation) SetError(fieldName string, errMsg string) *Error {
	err := &Error{Key: fieldName, Field: fieldName, Tmpl: errMsg, Message: errMsg}
	v.setError(err)
	return err
}
```
How about changing the condition to `err == nil`?
```suggestion
		if err == nil {
			log.Warning(fmt.Sprintf("Set validation error failed: %v", err))
		}
```

Reply:
Ahh!! I forget to double check inside of it 😅

Reply:
https://github.com/Lahphim/go-crawler-challenge/pull/39/commits/d9db03a262c5968415abf7c65dc31ff8d37bed9a